readme.txt
Readme
-------------
First, using Makefile by type: make

Client part:
To test Client.c on normal website:
cmd line: ./client http://www.csc.uvic.ca/index.htm

To Server.c 200 ok
Create a folder named : web
 and creat a txt file named: haha.txt 

cmd line on server side: ./server 8070 web/
cmd line on client side: ./client 10.10.1.100:8070/haha.txt

It will shows the message in haha.txt file.


To test 404
cmd line on server side: ./server 8071 web/
cmd line on client side: ./client 10.10.1.100:8070/hha.txt

To test 501
Modify client.c and find "GET " Then change the line HTTP/1.0 to something else ex:HTTP/1.5
and use Makefile again.

cmd line on server side: ./server 8072 web/
cmd line on client side: ./client 10.10.1.100:8072/haha.txt


changes.txt
Changes:
change server.c
I used fscanf(file,"%ls",content); which only gives me one line of the html file
now i am using fread(content, MAX_REC_LEN, 1, file); which gives me full content.


client.c
/*------------------------------
* client.c
* Description: HTTP client program
* CSC 361
* Instructor: Kui Wu
-------------------------------*/

/* define maximal string and reply length, this is just an example.*/
/* MAX_RES_LEN should be defined larger (e.g. 4096) in real testing. */
#define MAX_STR_LEN 120
#define MAX_RES_LEN 120
#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>
#include <stdio.h>
#include<string.h>

/* --------- Main() routine ------------
 * three main task will be excuted:
 * accept the input URI and parse it into fragments for further operation
 * open socket connection with specified sockid ID
 * use the socket id to connect sopecified server
 * don't forget to handle errors
 */

main(int argc, char **argv)
{
    char uri[MAX_STR_LEN];
    char hostname[MAX_STR_LEN];
    char identifier[MAX_STR_LEN];
    int sockid, port;
    

   // printf("Open URI:  ");
   // scanf("%s", uri);

	strcpy(uri,argv[1]);
	//uri=argv[1];
	printf("web: %s",uri);
    parse_URI(uri, hostname, &port, identifier);
	//printf("ip1 = \"%s\"\n", identifier);;

    sockid = open_connection(hostname, port);
	
    perform_http(sockid, uri,hostname,identifier);
}

/*------ Parse an "uri" into "hostname" and resource "identifier" --------*/

parse_URI(char *uri, char *hostname, int *port, char *identifier)
{
	int succ= 0;
  		*port=80;

      if (sscanf(uri, "http://%99[^:]:%i/%199[^\n]", hostname, port, identifier) == 3) { succ = 1;}
      else if (sscanf(uri, "http://%99[^/]/%199[^\n]", hostname, identifier) == 2) { succ = 1;}
      else if (sscanf(uri, "%99[^:]:%i/%199[^\n]", hostname, port, identifier) == 3) { succ = 1;}
      else if (sscanf(uri, "%99[^/]/%199[^\n]", hostname, identifier) == 2) { succ = 1;}
	 if (succ!=1) {
		printf("Error. \n"); 
}
	//printf("ip1 = \"%s\"\n", hostname);
}	
/*http://www.google.ca/index.html

/* http://www.csc.uvic.ca/index.htm
/*------------------------------------*
* connect to a HTTP server using hostname and port,
* and get the resource specified by identifier
*--------------------------------------*/

perform_http(int sockid, char *uri, char *host, char *identifier)
{	
	 char recvline[4096];
	char haha[MAX_STR_LEN]="GET ";
	int bytes;
	/* connect to server and retrieve response */
 	printf("\n---Request begin---\n");
	strcat(haha,"http://");
	strcat(haha,host);
	strcat(haha,"/");
	strcat(haha,identifier);
	strcat(haha," HTTP/1.0\r\n\r\n");
	
	printf("%s",haha);
	printf("Host: ");
	printf("%s",host);
	bzero( recvline, sizeof recvline);
	//printf("%s",haha);
	write(sockid,haha,strlen(haha)+1);
	read(sockid,recvline, 4096);
	printf("\nConnection: Keep-Alive\n");
	printf("\n---Request end---\n");
	printf("HTTP request sent, awaiting response...\n");
	printf("\n---Response header---\n");
	printf("%s",recvline);
	//printf("\n--- Response body---\n");
	//printf("%s",recvline);
 	 //close(sockid,identifier);
}
/*---------------------------------------------------------------------------*
 *
 * open_conn() routine. It connects to a remote server on a specified port.
 *
 *---------------------------------------------------------------------------*/
int open_connection(char *hostname, int port)
{
	struct sockaddr_in serv_addr;

	struct hostent *server_ent;
  	int sockfd;
	
	sockfd=socket(AF_INET,SOCK_STREAM,0);
	if (sockfd < 0) error("ERROR opening socket");
	bzero(&serv_addr,sizeof serv_addr);
	server_ent= gethostbyname(hostname);
	memcpy(&serv_addr.sin_addr, server_ent->h_addr, server_ent->h_length);
	
	serv_addr.sin_family=AF_INET;
	serv_addr.sin_port=htons(port);
	
	
	
	connect(sockfd,(struct sockaddr*)&serv_addr,sizeof(serv_addr));
		// error("ERROR connecting");

  /* generate socket
   * connect socket to the host address
   */	
  return sockfd;
}


server.c
/*------------------------------
* server.c
* Description: HTTP server program
* CSC 361
* Instructor: Kui Wu
-------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <netdb.h>
#include <netinet/in.h>

#include <string.h>
#define MAX_REC_LEN 4096
#define MAX_STR_LEN 120         /* maximum string length */
#define SERVER_PORT_ID 9898     /* server port number */

void cleanExit();

/*---------------------main() routine--------------------------*
 * tasks for main
 * generate socket and get socket id,
 * max number of connection is 3 (maximum length the queue of pending connections may grow to)
 * Accept request from client and generate new socket
 * Communicate with client and close new socket after done
 *---------------------------------------------------------------------------*/
	struct sockaddr_in serv_addr, cli_addr;
	 char buffer[256]; 
         int newsockfd,portno,reallynewsockfd,n,clilen; /* return value of the accept() call */
	char address[MAX_STR_LEN];
main(int argc, char **argv)
{	
	
	strcpy(address,argv[2]);
	  newsockfd = socket(AF_INET, SOCK_STREAM, 0);
 	  if (newsockfd < 0) {
     		 perror("ERROR opening socket");
      		exit(1);
	   }
	 
	 bzero((char *) &serv_addr, sizeof(serv_addr));
	
	int a = atoi(argv[1]);
	if(argc > 2){
  	 portno = a;
   	}else if(argc==2){
	portno=80;	
	}
	//printf("checkkkkkk %d",portno);
  	 serv_addr.sin_family = AF_INET;
   	serv_addr.sin_addr.s_addr = INADDR_ANY;
   	serv_addr.sin_port = htons(portno);
 	   bind(newsockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr));
      		
	listen(newsockfd, 3);
	//if (listen(newsockfd, 3) == -1) {
            //perror("listen error");
          //  exit(1);
      //  }
//while(1){
perform_http(newsockfd,address);
//}
	
 }

/*---------------------------------------------------------------------------*
 *
 * cleans up opened sockets whe n killed by a signal.
 *
 *---------------------------------------------------------------------------*/

void cleanExit()
{
    exit(0);
}

/*---------------------------------------------------------------------------*
 *
 * Accepts a request from "sockid" and sends a response to "sockid".
 *
 *---------------------------------------------------------------------------*/

perform_http(int sockid)
{	
	int succ= 0;
	char cmd[120];
	char hostname[MAX_STR_LEN];
	char identifier[MAX_STR_LEN];
	char http[MAX_STR_LEN];
	char content[4096];
	char header[999]="HTTP/1.0 200 OK\n";
	 time_t t = time(0);
  	  struct tm *tm = localtime(&t);
   	 //printf("%s\n", asctime(tm));
	reallynewsockfd= accept(sockid, (struct sockaddr *)&cli_addr, &clilen);
	bzero(buffer,256);
  	 read( reallynewsockfd,buffer,255 );
	//printf("buffer is :%s",buffer);
	if (sscanf(buffer, "%s http://%[^/]%[^ ] %[^\n]", cmd, hostname, identifier,http) == 4) { succ = 1;}
	
	//printf("ip = \"%s\"\n", hostname);
	//printf("cmd = \"%s\"\n", cmd);
	//int y =strcmp("HTTP/1.0\r",http);
	//printf("Number = \%d\n", y);
	while (strcmp("HTTP/1.0\r",http) != 0){
	write(reallynewsockfd,"HTTP/1.0 501 Not Implemented\n",29);
	close(reallynewsockfd);
	return;
	}
	strcat(address, identifier);
	//printf("address is = \"%s\"\n", address);
	FILE *file;//opening file herererer.
	file = fopen(address, "r");
	if (file==NULL) {
	write(reallynewsockfd,"HTTP/1.0 404 Not Found\n",29);
	close(reallynewsockfd);
	return;
	}
	fseek(file, 0, SEEK_SET);
	//fgets(content, MAX_REC_LEN, file);
	fread(content, MAX_REC_LEN, 1, file);
	//fscanf(file,"%ls",content);
	fclose(file);
	//strcat(header, "HTTP/1.0 200 OK\n");
	strcat(header, "Data: ");
	strcat(header, asctime(tm));
	strcat(header, "\n");
	strcat(header, "--- Response body---\n");
	strcat(header,content);
	strcat(header, "\n");
	write(reallynewsockfd,header,1000);
	
	//printf(content);
	//write(reallynewsockfd,content,1000);
	//char haha[7]="hellop";
      	//write(reallynewsockfd,"hello world",12);
     	  close(reallynewsockfd);
}















